examples/Components/Typeahead/Typeahead.purs

```diff
@@ -2,178 +2,247 @@ module Docs.Components.Typeahead where
+  make simpler ones like dropdowns. It just comes down to how much of Select you
+  want to extend.
+
+  In this case, we'll extend the component so that it can manage a list of selections,
+  where no selected item should be displayed in the list of available options. We'll
+  need to extend the component state with the list of selections; we'll need to extend
+  the query algebra with the ability to click to remove items; and we'll go ahead and
+  add a new message that lets a parent component know when an item was removed.
+
+  Plus, we'll get fancy: we'll have a dropdown running *inside* Select, using our
+  previous work.
+-}
+
+-----
+-- Component
+
+-- Just like the dropdown, all we have to do to export a full component is just 
+-- specialize the types and supply a render function, query handler, and message 
+-- handler.
+component :: forall m. MonadAff m => H.Component HH.HTML Query Input Message m
+component = Select.component render handleQuery handleMessage
+
+
+-----
+-- State
+
+-- Select doesn't maintain any selections on its own, but we can extend it to 
+-- hold a list of selected items by adding new contents to state. We'll need
+-- three lists: a list of all items, a list of non-selected items, and a list
+-- of selected items.
+type State = Select.State String ExtraState
+
+type ExtraState =
+  ( selectedItems :: Array String
+  , visibleItems :: Array String
+  )
+
+-- Our input type will also need these items
+type Input = Select.Input String ExtraState
+
+
+-----
+-- Child Components
+
+-- We're going to embed a dropdown in the component, so we'll extend Select with
+-- new child slots.
+type ChildSlots = 
+  ( dropdown :: H.Slot Dropdown.Query Dropdown.Message Unit )
+
+_dropdown = SProxy :: SProxy "dropdown"
+
+
+-----
+-- Messages
+
+-- We'll add a new message to Select to cover the case when an item is removed. The
+-- parent might wish to display a confirmation of the last-removed item or use it
+-- otherwise.
+type Message = Select.Message String ExtraMessage
+
+data ExtraMessage 
+  = ItemRemoved String 
+
+-- We also need to take action when an item is selected in Select or a debounced
+-- search has completed. We'll do that by handling the message within Select.
+handleMessage 
+  :: forall m
+   . MonadAff m
+  => Message 
+  -> H.HalogenM State Action ChildSlots Message m Unit
+handleMessage = case _ of
+  Select.Selected item -> do
+    st <- H.get
+    let newSelections = item : st.selectedItems
+    H.modify_ _ 
+      { selectedItems = newSelections
+      , visibleItems = difference newSelections st.items
+      , search = ""
+      }
+  
+  Select.Searched str -> do
+    st <- H.get
+    let 
+      visible = difference st.selectedItems $ filter (contains (Pattern str)) st.items
+      index = elemIndex str visible
+    H.modify_ _ { visibleItems = visible }
+    for_ index \ix -> do
+      Select.handleAction handleQuery handleMessage $ Select.Highlight $ Select.Index ix
```

Comment:
Feels a bit clunky having to pass these along. Though I guess if users find it annoying they can create their own helper
```purescript
handleAction = Select.handleAction handleQuery handleMessage
```

Reply:
Can't do that ^ -- unfortunately it leads to infinite recursion if you don't eta-expand. i don't know why. i'm probably going to file a ticket.